File: Code Wars Simple Encryption 1 Alternating Split.py
"""
Notes:
In the encrypt method, making char_array either a list or string
does not affect the output. However, it was determined that
it would be better to make it a list instead since an str instance
has 76 methods as of python 3.4 versus a list instance' 45.
The assumption is that the less methods, the less memory space used.
It is assumed that processing speed is the same for either instance
type.

Called len(dir(method)) on both str and list.
"""


def decrypt(encrypted_text, n):
    if not (isinstance(encrypted_text, str)):
        return encrypted_text

    char_array = [char for char in encrypted_text]
    integerdiv_half = len(char_array) // 2
    has_extra = False
    iterations = 0

    if (integerdiv_half * 2) < len(char_array):
        has_extra = True

    print("Input to Decrypt:", encrypted_text, "Iter:", n, sep='')
    print(
        "Integer division by two:", integerdiv_half,
        "Has extra?:", has_extra)

    while(iterations < n):
        firsts, seconds = [], []

        seconds = char_array[0: integerdiv_half]
        firsts = char_array[integerdiv_half: (integerdiv_half * 2)]

        if has_extra:
            firsts += [char_array[-1]]

        # print("Firsts:", ''.join(firsts), "Seconds:", ''.join(seconds))

        char_array.clear()

        for index in range(integerdiv_half):
            char_array.append(firsts[index])
            # print(index, ':', firsts[index])

            char_array.append(seconds[index])
            # print(index, ':', seconds[index])

        # char_array += firsts
        # char_array += seconds

        if has_extra:
            char_array += firsts[-1]

        # print("Char array at iter", iterations, "\n:", char_array)
        iterations += 1

    return ''.join(char_array)


def encrypt(text, n):
    if not (isinstance(text, str)):
        return text

    char_array = [char for char in text]
    iterations = 0

    print("Input to Encrypt:`", text, "`. Iter:", n, sep='')

    while(iterations < n):
        firsts, seconds = [], []

        for index in range(len(char_array)):
            if (index & 1 == 0):
                firsts += char_array[index]
            else:
                seconds += char_array[index]

        iterations += 1
        # print('\nSecs:', ''.join(seconds), "\nFirs:", ''.join(firsts))

        char_array.clear()

        char_array += (''.join(seconds))
        char_array += (''.join(firsts))

    return ''.join(char_array)


def test_input(dict_inputs, iterations=5):
    dict_results = {}

    for input_alias, string in dict_inputs.items():
        encrypted_text = encrypt(string, n=iterations)
        decrypted_text = decrypt(encrypted_text, n=iterations)
        dict_results[input_alias] = [string, encrypted_text, decrypted_text]

        print(
            """
            \n* Results for Input Alias: "{}".
            \nIterations:{}
            \nOriginal:\n\t"{}"
            \nEncrypted:\n\t"{}"
            \nDecrypted:\n\t"{}".""".format(
                    input_alias, iterations, string,
                    encrypted_text, decrypted_text
                    )
            )

    return dict_results


even_input = "This is even! End."
odd_input = "This is odd. End."

input_dict = {
    "even": even_input,
    "odd": odd_input
    }

test_input(input_dict)

test_input(input_dict, 8)

File: Code Wars Simple Encryption 1 Alternating Split.jl
"""
Decrypt method for Julia has apparently wrong steps but still works.
Conversely, the copy of the Python script section gives wrong innput.
- Tried switching to while-loop, same results.
- Tried switching the order of pushing char from firsts and seconds.
Gave wrong answer.
- Set iterations_done = 1 intead of zero then using python algorithm.
Python steps only worked on second test using 5 iterations and odd input.
Steps that shouldn't work still gave correct input.

Julia script appends! or pushes! all the seconds then all the firsts.
Proper procedure is to push in a first, push in a second recursively.
Encryption:
1. Group letters into pairs. From the pairs, distinguish firsts and seconds.
2. Put all second letters in output string.
3. Append all first letters to the output string.
4. Repeat for specified iterations.
"""

function encrypt(str::String, n::Int)
    text_length = length(str)
    arr_encrypted_text = [char for char in str]

    for ctr in 1:n        
        firsts, seconds = [], []

        for index in 1:text_length
            
            if (index & 1) == 1
                push!(firsts, arr_encrypted_text[index])
            else
                push!(seconds, arr_encrypted_text[index])
            end

        end
        
        empty!(arr_encrypted_text)        
        #append!(arr_encrypted_text, seconds, firsts)
        append!(arr_encrypted_text, seconds)
        append!(arr_encrypted_text, firsts)
        
        #for char in seconds
         #   push!(arr_encrypted_text, char)
        #end

        #for char in firsts
         #   push!(arr_encrypted_text, char)
        #end

    end

    return join(arr_encrypted_text)

end


function decrypt(str::String, n::Int)
    arr_text = [char for char in str]
    text_length = length(str)
    floordiv_half = Int(floor(text_length/2))
    #iterations_done = 1

    #while iterations_done < n                
    for ctr in 1:n
        firsts = arr_text[(floordiv_half + 1):text_length]
        seconds = arr_text[1:floordiv_half]

        empty!(arr_text)
        append!(arr_text, seconds)
        append!(arr_text, firsts)
        
        #Method that should not work: add seconds then add firsts.
        """
        for char in seconds
            push!(arr_text, char)
        end

        for char in firsts
            push!(arr_text, char)
        end
        """   

        #Copy of method in Python script. Doesn't work though it should.
        """
        for index in 1:floordiv_half
            push!(arr_text, seconds[index])
            push!(arr_text, firsts[index])
        end

        if (floordiv_half * 2) < text_length
            push!(arr_text, firsts[floordiv_half + 1])
        end
        """
        #iterations_done += 1

    end

    return join(arr_text)

end


function test_input(lst_strs::Array{String}, iter::Int)
    
    for string in lst_strs
        #println("\n=====\nOriginal:\n\t$str")
        encrypted_text = encrypt(string, iter)
        decrypted_text = decrypt(string, iter)
        
        println("\n=====\nOriginal:\n\t$string")
        println("Encrypted:\n\t", encrypted_text)
        println("Decrypted:\n\t", decrypted_text)
    end
    
end


even_input = "This is even! End."
odd_input = "This is odd. End."

test_input([even_input, odd_input], 5)
test_input([even_input, odd_input], 8)
